Python/Product/VSInterpreters/StringListReader.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.PythonTools.Interpreter {
+    class StringListReader : TextReader {
+        private readonly IEnumerator<string> _strings;
+        private StringReader _current;
+
+        public StringListReader(IEnumerable<string> strings) {
+            _strings = strings.GetEnumerator();
+            if (_strings.MoveNext()) {
+                _current = new StringReader(_strings.Current);
+            }
+        }
+
+        protected override void Dispose(bool disposing) {
+            base.Dispose(disposing);
+            if (disposing) {
+                _strings.Dispose();
+            }
+        }
+
+        private bool Next() {
+            if (_current != null) {
+                if (_strings.MoveNext()) {
+                    _current = new StringReader(_strings.Current);
+                    return true;
+                } else {
+                    _current = null;
+                }
+            }
+            return false;
+        }
+
+        public override string ReadLine() {
+            var r = _current?.ReadLine();
+            if (r == null && Next()) {
+                return _current.ReadLine();
+            }
+            return r;
+        }
+
+        public override int Read() {
+            if (_current == null) {
+                return -1;
+            }
+
+            var i = _current.Read();
+            if (i < 0 && Next()) {
+                return _current.Read();
+            }
+            return i;
+        }
+
+        public override int Read(char[] buffer, int index, int count) {
+            if (_current == null) {
+                return 0;
+            }
+            var i = _current.Read(buffer, index, count);
+            if (i == 0 && Next()) {
```

Comment:
If the requested count requires reading from 2 lines, this won't do it. I've seen some stream implementations behaving this way (not filling the requested count, even though there was more data available), and that was unexpected.
Remarks at https://msdn.microsoft.com/en-us/library/97e90x91(v=vs.110).aspx seem to imply that it should always fill the requested count, except when reached EOF.

Reply:
I've always understood it to mean "read again", but it looks like `ReadBlock` is intended to have the semantics you describe. I'll update that one